pkg/reconciler/run/controller.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2020 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package run
+
+import (
+	"context"
+
+	"github.com/tektoncd/pipeline/pkg/apis/config"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline"
+	pipelineclient "github.com/tektoncd/pipeline/pkg/client/injection/client"
+	runinformer "github.com/tektoncd/pipeline/pkg/client/injection/informers/pipeline/v1alpha1/run"
+	runreconciler "github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1alpha1/run"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/tracker"
+)
+
+// NewController instantiates a new controller.Impl from knative.dev/pkg/controller
+func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
```

Comment:
I am still trying to understand how controller initialization is happening in general. Here `NewController` is returning   implementation of the controller whereas `taskRun` and `pipelineRun` controllers are returning function:

https://github.com/tektoncd/pipeline/blob/732d2363b9df0b5c2e6b1de732a819b69c70bb3f/pkg/reconciler/taskrun/controller.go#L45-L46

https://github.com/tektoncd/pipeline/blob/732d2363b9df0b5c2e6b1de732a819b69c70bb3f/pkg/reconciler/pipelinerun/controller.go#L45-L46

Why does this defer from the `pipelineRun` and `taskRun`?

Reply:
Good point. I think we may want to follow the same pattern. In PipelineRun and TaskRun reconciler, this is mainly because we need to pass arguments (and the func return to satisfy the knative/pkg controller contract). We should support at least the `namespace` for `Run` too I guess ? (I don't think we need to pass the `images`)

Reply:
I had considered supporting namespaced-reconciliation for Runs in this PR but thought it would be simpler not to, to keep this change small. If you think it makes it easier to read here, though, I can do that.

Reply:
Yup would be great to exercise the same contract and easier to follow. If it cannot be part of this PR, at least a comment would be helpful 🙏